keps/sig-node/3545-improved-multi-numa-alignment/README.md

```diff
@@ -0,0 +1,483 @@
+
+This limitation surfaces in multi-socket, as well as single-socket multi NUMA systems (commonly present in AMD hardware), and can cause significant performance degradation in latency-critical execution and high-throughput applications if the `TopologyManager` decides to align resources on non-adjacent NUMA nodes.
+
+
+### Goals
+
+* Enable `TopologyManager` to prefer sets of NUMA nodes with shorter NUMA distances for all TopologyManager policies
+* Preserve all other properties of all policies
+
+## Proposal
+
+### Risks and Mitigations
+
+| Risk                                             | Impact | Mitigation                                                                                                                                                                               |
+|--------------------------------------------------|--------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| Bugs in the implementation lead to kubelet crash | High   | Disable the policy option and restart the kubelet. The workload will run but resources allocation can be spread among NUMA nodes which are non-adjacent what can lead to higher latency. |
+
+## Design Details
+
+
+### Proposed Change
+ 
+We propose to
+- add a new flag in Kubelet called `TopologyManagerPolicyOptions` in the kubelet config or command line argument called `topology-manager-policy-options` which allows the user to specify the Topology Manager policy option.
+- add a new topology manager option called `prefer-closest-numa-nodes`; if present, this option will enable further refinements of the existing `restricted` and `best-effort` policies, this option has no effect for `none` and `single-numa-node` policies.
+
+Best-effort or restricted policy chooses resources that will fit into the smallest number of NUMA nodes.
+This enhancement wants to change how the best hint is chosen from choosing narrower `NUMANodeAffinity` bitmask to choosing the one that is the narrowest and has the shortest average distance between NUMA nodes.
+
+To summarize properties of `prefer-closest-numa-nodes` policy:
+
+* Preserve all properties of `best-effort` and `restricted` policies
+* Choose `NUMANodeAffinity` bitmask which is the narrowest and if the bitmask width is the same prefer the bitmask with smaller average distance.
+
+When `prefer-closest-numa-nodes` policy is enabled, we need to retrieve information regarding distances between NUMA nodes.
+Right now Topology manager discovers Node layout using [CAdvisor API](https://github.com/google/cadvisor/blob/master/info/v1/machine.go#L40).
+
+We will need to extend the `MachineInfo` struct with a `Distances` field which will describe the distance between a given NUMA node and other NUMA nodes in the system.
+This information can be read from sysfs:
```

Comment:
Now that https://github.com/google/cadvisor/pull/3179 has been merged, can we remove references to the workaround (of obtaining distance information from `sysfs`)? Does the release cadence of cadvisor allow us to pull in the latest changes in Kubernetes 1.26 or we anticipate some lag and are intentionally keeping the workaround still here?

Reply:
This is merged and available on `master` branch but not in any of official releases. I am not familiar with how `cadvisor` release schedule and I can't guarantee that this change will be released before Kubernetes `v1.26` that is why I have kept this workaround here.

Reply:
Sure, thanks for clarifying. We can keep an eye on cadvisor release and adapt accordingly during the development phase depending on whether or not it makes its way into Kubernetes 1.26.